From 122ad46137ec1fae695b8047c5ae1eafef0a7491 Mon Sep 17 00:00:00 2001
From: Andrew Lukoshko <alukoshko@almalinux.org>
Date: Thu, 8 Aug 2024 16:36:33 +0000
Subject: [PATCH 01/10] Enable all disabled pci devices by moving to
 unmaintained list

---
 kernel/rh_messages.h | 66 ++++++++++++++++++++++----------------------
 1 file changed, 33 insertions(+), 33 deletions(-)

diff --git a/kernel/rh_messages.h b/kernel/rh_messages.h
index 3ac7e00e0..c0737616b 100644
--- a/kernel/rh_messages.h
+++ b/kernel/rh_messages.h
@@ -123,6 +123,39 @@ static const struct pci_device_id rh_deprecated_pci_devices[] = {
 };
 
 static const struct pci_device_id rh_disabled_pci_devices[] = {
+	{0}     /* Terminating entry */
+};
+
+static const struct pci_device_id rh_unmaintained_pci_devices[] = {
+	{ 0x10df, 0xe220, PCI_ANY_ID, PCI_ANY_ID },
+	{ 0x10df, 0x0724, PCI_ANY_ID, PCI_ANY_ID },
+	{ 0x10df, 0xe200, PCI_ANY_ID, PCI_ANY_ID },
+	{ 0x10df, 0xf011, PCI_ANY_ID, PCI_ANY_ID },
+	{ 0x10df, 0xf015, PCI_ANY_ID, PCI_ANY_ID },
+	{ 0x10df, 0xf100, PCI_ANY_ID, PCI_ANY_ID },
+	{ 0x10df, 0xfc40, PCI_ANY_ID, PCI_ANY_ID },
+	{ 0x1000, 0x005b, PCI_ANY_ID, PCI_ANY_ID },
+	{ 0x1000, 0x0071, PCI_ANY_ID, PCI_ANY_ID },
+	{ 0x1000, 0x0073, PCI_ANY_ID, PCI_ANY_ID },
+	{ 0x1000, 0x0079, PCI_ANY_ID, PCI_ANY_ID },
+	{ 0x15B3, 0xA2DC, PCI_ANY_ID, PCI_ANY_ID },
+	{ 0x1000, 0x006E, PCI_ANY_ID, PCI_ANY_ID },
+	{ 0x1000, 0x0080, PCI_ANY_ID, PCI_ANY_ID },
+	{ 0x1000, 0x0081, PCI_ANY_ID, PCI_ANY_ID },
+	{ 0x1000, 0x0082, PCI_ANY_ID, PCI_ANY_ID },
+	{ 0x1000, 0x0083, PCI_ANY_ID, PCI_ANY_ID },
+	{ 0x1000, 0x0084, PCI_ANY_ID, PCI_ANY_ID },
+	{ 0x1000, 0x0085, PCI_ANY_ID, PCI_ANY_ID },
+	{ 0x1000, 0x0086, PCI_ANY_ID, PCI_ANY_ID },
+	{ 0x1000, 0x0087, PCI_ANY_ID, PCI_ANY_ID },
+	{ 0x177d, 0xa01e, PCI_ANY_ID, PCI_ANY_ID },
+	{ 0x177d, 0xa034, PCI_ANY_ID, PCI_ANY_ID },
+	{ 0x177d, 0x0011, PCI_ANY_ID, PCI_ANY_ID },
+	{ 0x1077, 0x2031, PCI_ANY_ID, PCI_ANY_ID },
+	{ 0x1077, 0x2532, PCI_ANY_ID, PCI_ANY_ID },
+	{ 0x1077, 0x8031, PCI_ANY_ID, PCI_ANY_ID },
+	{ 0x1924, 0x0803, PCI_ANY_ID, PCI_ANY_ID },
+	{ 0x1924, 0x0813, PCI_ANY_ID, PCI_ANY_ID },
 	{ 0x1011, 0x0046, 0x103c, 0x10c2 },
 	{ 0x1011, 0x0046, 0x9005, 0x0364 },
 	{ 0x1011, 0x0046, 0x9005, 0x0365 },
@@ -289,37 +322,4 @@ static const struct pci_device_id rh_disabled_pci_devices[] = {
 	{0}     /* Terminating entry */
 };
 
-static const struct pci_device_id rh_unmaintained_pci_devices[] = {
-	{ 0x10df, 0xe220, PCI_ANY_ID, PCI_ANY_ID },
-	{ 0x10df, 0x0724, PCI_ANY_ID, PCI_ANY_ID },
-	{ 0x10df, 0xe200, PCI_ANY_ID, PCI_ANY_ID },
-	{ 0x10df, 0xf011, PCI_ANY_ID, PCI_ANY_ID },
-	{ 0x10df, 0xf015, PCI_ANY_ID, PCI_ANY_ID },
-	{ 0x10df, 0xf100, PCI_ANY_ID, PCI_ANY_ID },
-	{ 0x10df, 0xfc40, PCI_ANY_ID, PCI_ANY_ID },
-	{ 0x1000, 0x005b, PCI_ANY_ID, PCI_ANY_ID },
-	{ 0x1000, 0x0071, PCI_ANY_ID, PCI_ANY_ID },
-	{ 0x1000, 0x0073, PCI_ANY_ID, PCI_ANY_ID },
-	{ 0x1000, 0x0079, PCI_ANY_ID, PCI_ANY_ID },
-	{ 0x15B3, 0xA2DC, PCI_ANY_ID, PCI_ANY_ID },
-	{ 0x1000, 0x006E, PCI_ANY_ID, PCI_ANY_ID },
-	{ 0x1000, 0x0080, PCI_ANY_ID, PCI_ANY_ID },
-	{ 0x1000, 0x0081, PCI_ANY_ID, PCI_ANY_ID },
-	{ 0x1000, 0x0082, PCI_ANY_ID, PCI_ANY_ID },
-	{ 0x1000, 0x0083, PCI_ANY_ID, PCI_ANY_ID },
-	{ 0x1000, 0x0084, PCI_ANY_ID, PCI_ANY_ID },
-	{ 0x1000, 0x0085, PCI_ANY_ID, PCI_ANY_ID },
-	{ 0x1000, 0x0086, PCI_ANY_ID, PCI_ANY_ID },
-	{ 0x1000, 0x0087, PCI_ANY_ID, PCI_ANY_ID },
-	{ 0x177d, 0xa01e, PCI_ANY_ID, PCI_ANY_ID },
-	{ 0x177d, 0xa034, PCI_ANY_ID, PCI_ANY_ID },
-	{ 0x177d, 0x0011, PCI_ANY_ID, PCI_ANY_ID },
-	{ 0x1077, 0x2031, PCI_ANY_ID, PCI_ANY_ID },
-	{ 0x1077, 0x2532, PCI_ANY_ID, PCI_ANY_ID },
-	{ 0x1077, 0x8031, PCI_ANY_ID, PCI_ANY_ID },
-	{ 0x1924, 0x0803, PCI_ANY_ID, PCI_ANY_ID },
-	{ 0x1924, 0x0813, PCI_ANY_ID, PCI_ANY_ID },
-	{0}     /* Terminating entry */
-};
-
 #endif /* __RH_MESSAGES_H */
-- 
2.43.5

