From 3d04d890de1c000aa734a72ff62bc52bf6302a4b Mon Sep 17 00:00:00 2001
From: Xiong Zhang <xiong.y.zhang@intel.com>
Date: Wed, 19 Apr 2017 08:51:07 +0800
Subject: [PATCH 07/45] i915/gvt/xengt.c: New interface to write protect PPGTT

Signed-off-by: Xiong Zhang <xiong.y.zhang@intel.com>
---
 drivers/gpu/drm/i915/gvt/xengt.c | 62 +++++++++++++++++++++++-----------------
 1 file changed, 36 insertions(+), 26 deletions(-)

diff --git a/drivers/gpu/drm/i915/gvt/xengt.c b/drivers/gpu/drm/i915/gvt/xengt.c
index 22b4ee9..2325b69 100644
--- a/drivers/gpu/drm/i915/gvt/xengt.c
+++ b/drivers/gpu/drm/i915/gvt/xengt.c
@@ -651,6 +651,33 @@ static int xen_hvm_map_pcidev_to_ioreq_server(struct xengt_hvm_dev *info,
 	return r;
 }
 
+static int hvm_claim_ioreq_server_type(struct xengt_hvm_dev *info,
+		uint32_t set)
+{
+
+	xen_dm_op_buf_t dm_buf;
+	struct xen_dm_op op;
+	struct xen_dm_op_map_mem_type_to_ioreq_server *data;
+	int r;
+
+	memset(&op, 0, sizeof(op));
+
+	op.op = XEN_DMOP_map_mem_type_to_ioreq_server;
+	data = &op.u.map_mem_type_to_ioreq_server;
+	data->id = info->iosrv_id;
+	data->type = HVMMEM_ioreq_server;
+	data->flags = (set == 1) ? XEN_DMOP_IOREQ_MEM_ACCESS_WRITE : 0;
+
+	dm_buf.h = &op;
+	dm_buf.size = sizeof(op);
+
+	r = HYPERVISOR_dm_op(info->vm_id, 1, &dm_buf);
+	if (r < 0)
+		gvt_err("Cannot map mem type to ioreq_server\n");
+
+	return r;
+}
+
 static int xen_hvm_set_mem_type(domid_t vm_id, uint16_t mem_type,
 		uint64_t first_pfn, uint64_t nr)
 {
@@ -683,36 +710,13 @@ static int xen_hvm_wp_page_to_ioreq_server(struct xengt_hvm_dev *info,
 		unsigned long page, bool set)
 {
 	int rc = 0;
-	uint64_t start, end;
 	uint16_t mem_type;
 
-	start = page << PAGE_SHIFT;
-	end = ((page + 1) << PAGE_SHIFT) - 1;
-
-	if (set) {
-		rc = xen_hvm_map_io_range_to_ioreq_server(info, 1,
-				start, end, true);
-		if (rc < 0) {
-			gvt_err("map page 0x%lx failed: %d!\n",	page, rc);
-			return rc;
-		}
-	}
-
-	mem_type = set ? HVMMEM_mmio_write_dm : HVMMEM_ram_rw;
+	mem_type = set ? HVMMEM_ioreq_server : HVMMEM_ram_rw;
 	rc = xen_hvm_set_mem_type(info->vm_id, mem_type, page, 1);
 	if (rc < 0) {
 		gvt_err("set mem type of page 0x%lx to %s fail - %d!\n", page,
-				set ? "HVMMEM_mmio_write_dm" : "HVMMEM_ram_rw", rc);
-		return rc;
-	}
-
-	if (!set) {
-		rc = xen_hvm_map_io_range_to_ioreq_server(info, 1,
-				start, end, false);
-		if (rc < 0) {
-			gvt_err("unmap page 0x%lx failed: %d!\n", page, rc);
-			return rc;
-		}
+				set ? "HVMMEM_ioreq_server" : "HVMMEM_ram_rw", rc);
 	}
 
 	return rc;
@@ -1269,8 +1273,10 @@ void xengt_instance_destroy(struct intel_vgpu *vgpu)
 	if (!info->nr_vcpu || info->evtchn_irq == NULL)
 		goto out1;
 
-	if (info->iosrv_id != 0)
+	if (info->iosrv_id != 0) {
+		hvm_claim_ioreq_server_type(info, 0);
 		xen_hvm_destroy_iorequest_server(info);
+	}
 
 	for (vcpu = 0; vcpu < info->nr_vcpu; vcpu++) {
 		if (info->evtchn_irq[vcpu] >= 0)
@@ -1338,6 +1344,10 @@ struct intel_vgpu *xengt_instance_create(domid_t vm_id,
 	if (rc < 0)
 		goto err;
 
+	rc = hvm_claim_ioreq_server_type(info, 1);
+	if (rc < 0)
+		goto err;
+
 	rc = xen_hvm_toggle_iorequest_server(info, 1);
 	if (rc < 0)
 		goto err;
-- 
2.7.4

