From 022f2762fd15c84431e8e0d79381539edaffcf9e Mon Sep 17 00:00:00 2001
From: Corentin Oparowski <corentin.oparowski@vates.tech>
Date: Sat, 19 Sep 2015 08:09:42 +0200
Subject: [PATCH 18/36] 0001-xen-swiotlb-size-128MiB.patch

---
 include/linux/swiotlb.h | 1 +
 kernel/dma/swiotlb.c    | 4 ++++
 2 files changed, 5 insertions(+)

diff --git a/include/linux/swiotlb.h b/include/linux/swiotlb.h
index 3dae0f592063e..f9a0bf692c9fd 100644
--- a/include/linux/swiotlb.h
+++ b/include/linux/swiotlb.h
@@ -34,6 +34,7 @@ struct scatterlist;
 
 /* default to 64MB */
 #define IO_TLB_DEFAULT_SIZE (64UL<<20)
+#define XEN_IO_TLB_DEFAULT_SIZE (128UL<<20)
 
 unsigned long swiotlb_size_or_default(void);
 void __init swiotlb_init_remap(bool addressing_limit, unsigned int flags,
diff --git a/kernel/dma/swiotlb.c b/kernel/dma/swiotlb.c
index e2445b3af7468..90c4778490add 100644
--- a/kernel/dma/swiotlb.c
+++ b/kernel/dma/swiotlb.c
@@ -100,7 +100,11 @@ static struct io_tlb_mem io_tlb_default_mem;
 
 #endif	/* CONFIG_SWIOTLB_DYNAMIC */
 
+#ifdef CONFIG_SWIOTLB_XEN
+static unsigned long default_nslabs = XEN_IO_TLB_DEFAULT_SIZE >> IO_TLB_SHIFT;
+#else
 static unsigned long default_nslabs = IO_TLB_DEFAULT_SIZE >> IO_TLB_SHIFT;
+#endif
 static unsigned long default_nareas;
 
 /**
-- 
2.47.3

