diff --git a/drivers/block/loop.c b/drivers/block/loop.c
index 80cf8ad..31cb386 100644
--- a/drivers/block/loop.c
+++ b/drivers/block/loop.c
@@ -1578,7 +1578,7 @@ static void lo_release(struct gendisk *disk, fmode_t mode)
 		err = loop_clr_fd(lo);
 		if (!err)
 			return;
-	} else {
+	} else if (lo->lo_state == Lo_bound) {
 		/*
 		 * Otherwise keep thread (if running) and config,
 		 * but flush possible ongoing bios in thread.
