diff --git a/arch/x86/xen/pci-swiotlb-xen.c b/arch/x86/xen/pci-swiotlb-xen.c
index b7df210..15064be 100644
--- a/arch/x86/xen/pci-swiotlb-xen.c
+++ b/arch/x86/xen/pci-swiotlb-xen.c
@@ -172,9 +172,6 @@ static int check_batch(int size)
 	return res;
 }
 
-
-
-
 static int pv_iommu_setup(void *data)
 {
 	int pfn, count = 0;
@@ -249,6 +246,11 @@ static int pv_iommu_setup(void *data)
 	return 0;
 }
 
+unsigned int xen_pv_iommu_is_ready(void)
+{
+    return pv_iommu_1_to_1_setup_complete;
+}
+
 /*
  * pci_xen_swiotlb_detect - set xen_swiotlb to 1 if necessary
  *
diff --git a/kernel/ksysfs.c b/kernel/ksysfs.c
index e83b264..2bfb05f 100644
--- a/kernel/ksysfs.c
+++ b/kernel/ksysfs.c
@@ -161,6 +161,15 @@ static ssize_t rcu_expedited_store(struct kobject *kobj,
 }
 KERNEL_ATTR_RW(rcu_expedited);
 
+extern unsigned int xen_pv_iommu_is_ready(void);
+
+static ssize_t pv_iommu_ready_show(struct kobject *kobj,
+				  struct kobj_attribute *attr, char *buf)
+{
+	return sprintf(buf, "%u\n", xen_pv_iommu_is_ready());
+}
+KERNEL_ATTR_RO(pv_iommu_ready);
+
 /*
  * Make /sys/kernel/notes give the raw contents of our kernel .notes section.
  */
@@ -203,6 +212,7 @@ static struct attribute * kernel_attrs[] = {
 	&vmcoreinfo_attr.attr,
 #endif
 	&rcu_expedited_attr.attr,
+	&pv_iommu_ready_attr.attr,
 	NULL
 };
 
