--- a/arch/x86/boot/main.c	2019-03-13 04:04:53.000000000 -0700
+++ b/arch/x86/boot/main.c	2019-05-25 14:31:21.043272496 -0700
@@ -147,7 +147,7 @@ void main(void)
 
 	/* Make sure we have all the proper CPU support */
 	if (validate_cpu()) {
-		puts("This processor is not supported in this version of RHEL.\n");
+		puts("This processor is not supported in this version of EuroLinux.\n");
 		die();
 	}
 

