diff --git a/drivers/firmware/efi/efi.c b/drivers/firmware/efi/efi.c
index 2a29dd9..687ed4f 100644
--- a/drivers/firmware/efi/efi.c
+++ b/drivers/firmware/efi/efi.c
@@ -575,7 +575,7 @@ int __init efi_config_parse_tables(void *config_tables, int count, int sz,
 	efi_tpm_eventlog_init();
 
 	/* Parse the EFI Properties table if it exists */
-	if (efi.properties_table != EFI_INVALID_TABLE_ADDR) {
+	if (efi.properties_table != EFI_INVALID_TABLE_ADDR && !xen_initial_domain()) {
 		efi_properties_table_t *tbl;
 
 		tbl = early_memremap(efi.properties_table, sizeof(*tbl));
