diff --git a/drivers/gpu/drm/i915/gvt/cfg_space.c b/drivers/gpu/drm/i915/gvt/cfg_space.c
index cd1dd3f..89d277e 100644
--- a/drivers/gpu/drm/i915/gvt/cfg_space.c
+++ b/drivers/gpu/drm/i915/gvt/cfg_space.c
@@ -358,6 +358,11 @@ void intel_vgpu_init_cfg_space(struct intel_vgpu *vgpu,
 			INTEL_GVT_PCI_CLASS_VGA_OTHER;
 		vgpu_cfg_space(vgpu)[PCI_CLASS_PROG] =
 			INTEL_GVT_PCI_CLASS_VGA_OTHER;
+	} else {
+		vgpu_cfg_space(vgpu)[PCI_CLASS_DEVICE] =
+			INTEL_GVT_PCI_CLASS_VGA_COMPAT;
+		vgpu_cfg_space(vgpu)[PCI_CLASS_PROG] =
+			INTEL_GVT_PCI_CLASS_VGA_COMPAT;
 	}
 
 	/* Show guest that there isn't any stolen memory.*/
diff --git a/drivers/gpu/drm/i915/gvt/reg.h b/drivers/gpu/drm/i915/gvt/reg.h
index 83f2f63..678db3f 100644
--- a/drivers/gpu/drm/i915/gvt/reg.h
+++ b/drivers/gpu/drm/i915/gvt/reg.h
@@ -25,6 +25,7 @@
 #define _GVT_REG_H
 
 #define INTEL_GVT_PCI_CLASS_VGA_OTHER   0x80
+#define INTEL_GVT_PCI_CLASS_VGA_COMPAT  0x00
 
 #define INTEL_GVT_PCI_GMCH_CONTROL	0x50
 #define   BDW_GMCH_GMS_SHIFT		8
