diff --git a/drivers/base/Kconfig b/drivers/base/Kconfig
index 98504ec..4f9bab9 100644
--- a/drivers/base/Kconfig
+++ b/drivers/base/Kconfig
@@ -229,7 +229,7 @@ config SOC_BUS
 source "drivers/base/regmap/Kconfig"
 
 config DMA_SHARED_BUFFER
-	bool
+	bool "DMA Shared Buffer Framework"
 	default n
 	select ANON_INODES
 	help
diff --git a/drivers/video/Kconfig b/drivers/video/Kconfig
index e0606c0..21447d1 100644
--- a/drivers/video/Kconfig
+++ b/drivers/video/Kconfig
@@ -38,7 +38,7 @@ config VIDEOMODE_HELPERS
 	bool
 
 config HDMI
-	bool
+	bool "Enable HDMI support"
 
 if VT
 	source "drivers/video/console/Kconfig"
diff --git a/lib/Kconfig b/lib/Kconfig
index 1a48744..0ce2799 100644
--- a/lib/Kconfig
+++ b/lib/Kconfig
@@ -349,7 +349,7 @@ config BTREE
 	bool
 
 config INTERVAL_TREE
-	bool
+	bool "Interval Tree"
 	help
 	  Simple, embeddable, interval-tree. Can find the start of an
 	  overlapping range in log(n) time and then iterate over all
