diff --git a/arch/x86/xen/setup.c b/arch/x86/xen/setup.c
index 7ab2951..0cc2312 100644
--- a/arch/x86/xen/setup.c
+++ b/arch/x86/xen/setup.c
@@ -794,8 +794,14 @@ char * __init xen_memory_setup(void)
 	 * UNUSABLE regions in domUs are not handled and will need
 	 * a patch in the future.
 	 */
-	if (xen_initial_domain())
+	if (xen_initial_domain()) {
 		xen_ignore_unusable();
+		/* Reserve 0.5 MiB to 1 MiB region so iBFT can be found */
+		xen_e820_map[xen_e820_map_entries].addr = 0x80000;
+		xen_e820_map[xen_e820_map_entries].size = 0x80000;
+		xen_e820_map[xen_e820_map_entries].type = E820_RESERVED;
+		xen_e820_map_entries++;
+	}
 
 	/* Make sure the Xen-supplied memory map is well-ordered. */
 	sanitize_e820_map(xen_e820_map, ARRAY_SIZE(xen_e820_map),
