diff --git a/include/xen/interface/memory.h b/include/xen/interface/memory.h
index 786d240..d0090d5 100644
--- a/include/xen/interface/memory.h
+++ b/include/xen/interface/memory.h
@@ -251,7 +251,7 @@ DEFINE_GUEST_HANDLE_STRUCT(xen_memory_map);
 /*
  * Translate the given guest PFNs to MFNs
  */
-#define XENMEM_get_mfn_from_pfn    25
+#define XENMEM_get_mfn_from_pfn    29
 struct xen_get_mfn_from_pfn {
     /*
      * Pointer to buffer to fill with list of pfn.
