---
 net/sunrpc/xprt.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/net/sunrpc/xprt.c b/net/sunrpc/xprt.c
index 8eb0c4f..bdd0162 100644
--- a/net/sunrpc/xprt.c
+++ b/net/sunrpc/xprt.c
@@ -811,6 +811,16 @@ static void xprt_connect_status(struct rpc_task *task)
 	case -ETIMEDOUT:
 		dprintk("RPC: %5u xprt_connect_status: connect attempt timed "
 				"out\n", task->tk_pid);
+		/*
+		 * Timed out while waiting for the old connection to
+		 * be closed?
+		 *
+		 * Force a disconnect since its unlikely to close
+		 * gracefully and waiting for the socket close to
+		 * timeout may take a long time.
+		 */
+		if (test_bit(XPRT_CLOSING, &xprt->state))
+			xprt_force_disconnect(xprt);
 		break;
 	default:
 		dprintk("RPC: %5u xprt_connect_status: error %d connecting to "
