diff --git a/drivers/gpu/drm/i915/gvt/aperture_gm.c b/drivers/gpu/drm/i915/gvt/aperture_gm.c
index 59c2681..61e7397 100644
--- a/drivers/gpu/drm/i915/gvt/aperture_gm.c
+++ b/drivers/gpu/drm/i915/gvt/aperture_gm.c
@@ -62,7 +62,7 @@ static int alloc_gm(struct intel_vgpu *vgpu, bool high_gm)
 
 	mutex_lock(&dev_priv->drm.struct_mutex);
 	ret = i915_gem_gtt_insert(&dev_priv->ggtt.vm, node,
-				  size, I915_GTT_PAGE_SIZE,
+				  size, I915_GTT_PAGE_SIZE_2M,
 				  I915_COLOR_UNEVICTABLE,
 				  start, end, flags);
 	mutex_unlock(&dev_priv->drm.struct_mutex);
diff --git a/drivers/gpu/drm/i915/gvt/xengt.c b/drivers/gpu/drm/i915/gvt/xengt.c
index 8b4da45..c4ab8d7 100644
--- a/drivers/gpu/drm/i915/gvt/xengt.c
+++ b/drivers/gpu/drm/i915/gvt/xengt.c
@@ -718,7 +718,7 @@ static int xen_hvm_memory_mapping(domid_t vm_id, uint64_t first_gfn,
 {
 	struct xen_domctl arg;
 	int rc = 0, err = 0;
-	unsigned long done = 0, mapping_sz = 64;
+	unsigned long done = 0, mapping_sz = nr_mfns;
 
 	if (add_mapping) {
 		rc = xen_domain_iomem_perm(vm_id, first_mfn, nr_mfns, 1);
@@ -745,6 +745,10 @@ static int xen_hvm_memory_mapping(domid_t vm_id, uint64_t first_gfn,
 				mapping_sz /= 2;
 				goto retry;
 			}
+			if (err > 0) {
+				done += err;
+				continue;
+			}
 			//Save first error status.
 			if (!rc)
 				rc = err;
